---
title: "Dunlop growth rate - internal standard normalization"
author: "Cassi Wattenburger"
date: "2/22/2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(results = "show")
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

**Goal of this script:** Normalize rarefied data by internal standard read count.

**Import libraries**

```{r}
rm(list=ls())

library("phyloseq")
library("ape")
library("vegan")
library("ggplot2")
library("reshape")
library("plyr")
library("dplyr")

sessionInfo()
```


**Import rarefied data**

```{r}
# Phyloseq objects created in 01_dunlopgr_rarefy.Rmd
load("rdata.files/gr_physeq.rarefy.RData")

# Pull data out of phyloseq
count <- data.frame(otu_table(physeq.rarefy))
tax <- data.frame(tax_table(physeq.rarefy))
meta <- data.frame(sample_data(physeq.rarefy))
```


**Identify internal standard ASV**

```{r}
# Find identifiers associated with spike-in genus Aquifex
tax.aquifex <- subset(tax, Genus=="Aquifex")
id.aquifex <- row.names(tax.aquifex)

# Create new row with summed spike-in counts
count.asv <- count
count.asv$ASV <- rownames(count)
count.spike <- filter(count.asv, ASV %in% id.aquifex)
rownames(count.spike) <- count.spike$ASV
count.spike <- count.spike[,-260]
count.spike <- as.matrix(count.spike)
spikesum <- colSums(count.spike)
spikesum <- as.data.frame(spikesum)
```

DADA2 seems to have split the internal standard into ~6 ASVs, but the read count abundance of the other 5 is minimal to none. Not sure what causes this behavior, but may be due to sequencing errors.


**Calculate proportion of internal standard per sample**

```{r}
# Calculate proportion of spike-in in each sample
totalsum <- colSums(count)
totalsum <- as.data.frame(totalsum)
totalspikesum <- merge(totalsum, spikesum, by=0)
totalspikesum <- mutate(totalspikesum, percent=(spikesum/totalsum)*100)

# Reformat
rownames(totalspikesum) <- totalspikesum[,1]
#totalspikesum$Row.names <- gsub("sa","", totalspikesum$Row.names)
#totalspikesum$Row.names <- as.numeric(totalspikesum$Row.names)
totalspikesum <- merge(meta, totalspikesum, by=0)
rownames(totalspikesum) <- totalspikesum[,1]
totalspikesum <- totalspikesum[,-c(1,12)]
```


**Add aggregated spike abundance back into data**

```{r}
# Add aggregated spike info back into count table
count.m <- as.matrix(count)
count.trans <- t(count.m)
spikesum$Sample <- rownames(spikesum)
count.trans.spike <- merge(count.trans, spikesum, by=0)
rownames(count.trans.spike) <- count.trans.spike[,1]
count.trans.spike <- count.trans.spike[,-1]
count.trans.spike <- count.trans.spike[,-33015]
```


**Assess relationship between % spike and DNA yield**

Should have negative relationship. The same amount of spike in more or less DNA should lead to a smaller or larger proportion of spike in the sample.

```{r}
# Add dna conc. metadata
dnaconc <- read.csv("growthrate_dnaext.csv")
totalspikesum <- merge(totalspikesum, dnaconc, by=1)
```

Spike percentage visualizations:

```{r}
# % standard in both experiments
ggplot(totalspikesum, aes(x=SampleID, y=percent)) +
  geom_point(aes(color=Cammend, shape=Soil)) +
  facet_wrap(~Experiment, scales="free") +
  scale_shape_manual(values=1:nlevels(totalspikesum$Soil)) +
  theme_test()

# % standard in ucosms only
ggplot(totalspikesum[totalspikesum$Experiment=="Microcosm",], aes(x=SampleID, y=percent)) +
  geom_point(aes(color=Cammend, shape=Soil)) +
  labs(title="Microcosm samples") +
  scale_shape_manual(values=1:nlevels(totalspikesum$Soil)) +
  theme_test() +
  theme(title=element_text(size=12),
      axis.text=element_text(size=12),
      axis.title=element_text(size=12)) +
  scale_y_continuous(limits=c(0,10))

# % standard vs dna yield
ggplot(totalspikesum, aes(x=Conc., y=log(percent))) +
  geom_point(aes(color=Cammend, shape=Soil)) +
  geom_smooth(method="lm", weight=0.5) +
  facet_wrap(~Experiment, scales="free") +
  scale_shape_manual(values=1:nlevels(totalspikesum$Soil)) +
  theme_test()
```

Strange gap in the samples. What's that about?

```{r}
ggplot(totalspikesum, aes(x=TimePoint, y=Conc.)) +
  geom_point(aes(color=Cammend, shape=Soil)) +
  geom_smooth(method="lm", weight=0.5) +
  #geom_text(aes(label=TimePoint), size=3) +
  scale_shape_manual(values=1:nlevels(totalspikesum$Soil)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme_test()

ggplot(totalspikesum, aes(x=Box, y=Conc.)) +
  geom_boxplot() +
    geom_jitter() +
  #geom_text(aes(label=TimePoint), size=3) +
  scale_shape_manual(values=1:nlevels(totalspikesum$Soil)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme_test()
```

Some boxes have higher concentration than others, but all have the weird gap.

**Statistics**

Is there a an effect of kit box used on DNA concentration?

```{r}
totalspikesum$Box <- as.character(totalspikesum$Box) # box number has no meaning other than as 
kruskal.test(Conc. ~ Box, data=totalspikesum)
# p-value = 4.362e-07

totalspikesum$Box <- as.factor(totalspikesum$Experiment)
kruskal.test(Conc. ~ Experiment, data=totalspikesum)
```

Unfortunately, I didn't randomize my DNA extractions, so this is confounded by samples processed in one kit being more similar to one another in other ways ie plot location, point in time series, treatment. It could even be who weighed out the soil that day.

Descriptive stats: Min, max, average, standard deviation of percent internal standard:

```{r}
t(data.frame(min = min(totalspikesum$percent),
           min.without.0s = min(totalspikesum[totalspikesum$percent > 0,]$percent),
           number.of.0s = nrow(totalspikesum[totalspikesum$percent == 0,]),
           max = max(totalspikesum$percent),
           avg = mean(totalspikesum$percent),
           stdev = sd(totalspikesum$percent)))
```

There was one sample that did not have any internal standard.

Is there a correlation between spike-in percentage and DNA concentration?

Hypothesis: negative correlation, more community DNA should result in a smaller proportion of the internal standard.

Spearman correlation (non-parametric), since variability looks greater at one end than the other.

```{r}
dnaspike.spear = cor.test(totalspikesum$Conc., log(totalspikesum$percent), method="spearman")
dnaspike.spear
```

Ties warning. Use permuted Spearman correlation instead.

See: https://stats.stackexchange.com/questions/50015/spearman-correlation-in-the-presence-of-many-ties-how-to-spot-a-problem

```{r}
# Create permuted spearman correlation function (test rho is less than 0)
spearman_permute = function(x, y) {
  set.seed(1) # make sure it is reproducible
  perm.estimates = c()
  for (i in 1:1000) { # permute 1000 times
    perm = sample(y) # randomize response variable
    perm.test = suppressWarnings(cor.test(x, perm, method="spearman", alternative="less")) # spearman correlation with permuted data
    perm.estimates = append(perm.estimates, perm.test$estimate) # permuted estimates
  }
  actual.test = suppressWarnings(cor.test(x, y, method="spearman", alternative="less")) # spearman correlation with actual data
  actual.estimate = actual.test$estimate # actual data estimate (rho)
  est.exceed = sum(abs(perm.estimates) < actual.estimate) # number of permutations that are less than the actual test rho
  est.equal = sum(abs(perm.estimates) == actual.estimate) # number of permutations that equal the actual test rho
  print(est.exceed)
  permuted.pval = (est.exceed + est.equal/2) / length(perm.estimates) # proportion of exceedances: the p-value
  return(c(permuted.pval, actual.estimate)) # the permuted p-value and estimate (rho)
}

# Test
dnaspike.spearperm = spearman_permute(totalspikesum$Conc., log(totalspikesum$percent)) # plotting as log, and relationship doesn't look linear
dnaspike.spearperm
```

Permuted spearman estimate agrees exactly with non-permuted.


**Normalize ASV abundances by internal standard**

Taxa reads / internal standard reads in the sample.

```{r}
# create dataframe to store normalized data
df.norm <- data.frame(matrix(NA, nrow=nrow(count.trans.spike), ncol=ncol(count.trans.spike)))
rownames(df.norm) <- row.names(count.trans.spike)
colnames(df.norm) <- colnames(count.trans.spike)

# Divide each taxa count (column) by corresponding spike count (spikesum column)
for (i in 1:ncol(count.trans.spike)) {
  df.norm[,i] <- count.trans.spike[,i]/count.trans.spike$spikesum
}
```

Reformat and save.

```{r}
# Add metadata to normalized data
count.meta.norm <- merge(meta, df.norm, by=0)

# Subset microcosm samples
microcosm.norm <- filter(count.meta.norm, Experiment %in% c("Microcosm"))
rownames(microcosm.norm) <- microcosm.norm[,2]

# Subset in-field samples
field.norm = filter(count.meta.norm, Experiment %in% c("Field"))
rownames(field.norm) <- field.norm[,2]
field.norm = field.norm[,-c(1,3:5,8:11)]
```

```{r}
# Reformat
microcosm.norm <- microcosm.norm[,-c(1,3:4,7:8,11)]
microcosm.norm.melt <- melt(microcosm.norm, id.vars=c("SampleID","Cammend","Soil","TimePoint","Days"))
field.norm.melt <- melt(field.norm, id.vars=c("SampleID","Soil","Block"))
```

```{r}
# Add taxonomy
tax$ASV <- rownames(tax)
microcosm.norm.tax <- merge(microcosm.norm.melt, tax, by.x=6, by.y=7)
field.norm.tax <- merge(field.norm.melt, tax, by.x=4, by.y=7)
```

```{r, eval=FALSE}
save(microcosm.norm.tax, file="rdata.files/gr_microcosm.norm.tax.RData")
save(field.norm.tax, file="rdata.files/gr_field.norm.tax.RData")
saveRDS(totalspikesum, file="rdata.files/gr_intstd.csv")
```


# Figures

```{r}
library(tidyverse)
library(cowplot)

totalspikesum <- readRDS(file="rdata.files/gr_intstd.csv")

# Clean up the dataframe for plotting
intstd = totalspikesum %>%
  select(Sample=SampleID, Soil, Amendment=Cammend, Block, Day=Days, Experiment, spikesum, totalsum, DNA_conc=Conc., Box) %>% # clean up
  filter(Soil %in% c("C3", "S17")) %>% # relevant soils
  mutate(spike_perc = (spikesum/totalsum)*100, Box = as.factor(Box)) # percent spike-in per sample

# Plot relationship with DNA conc.
p1 = intstd %>%
  ggplot(aes(x=DNA_conc, y=log(spike_perc))) +
  geom_point(size=1, shape=1) +
  labs(x="DNA concentration (ng/\u03BCL)", y="Internal standard (% of reads), ln") +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black"),
        axis.title.x = element_text(margin = unit(c(3, 0, 0, 0), "mm"), size=10),
        axis.title.y = element_text(margin = unit(c(0, 3, 0, 0), "mm"), size=10),
        axis.text = element_text(size = 8))
p1

# Box plot
p2 = intstd %>%
  ggplot(aes(x=DNA_conc, y=spike_perc)) +
  geom_boxplot() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        axis.text = element_blank(),
        axis.ticks=element_blank())
p2

#dnaspike.plot = plot_grid(p1, p2, align="h", axis="b", rel_widths=c(1, 0.1), labels=c("a","b"), label_size=10)
#dnaspike.plot
```

```{r, eval = FALSE}
library(Cairo)
#device=cairo_pdf,
ggsave(p1, file = "figures/suppfig_dnavsintstd.png", width = 90, height = 90, unit="mm")
```

