---
title: "Dunlop growth rates - 16S copy number vs start of growth"
author: "Cassandra Wattenburger"
date: "1/28/2021"
output: markdown_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(results = "show")
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

Goal of script: investigate relationship between 16S copy number and lag.

Hypothesis: Taxa with higher 16S copy number will have earlier start days.

This may depend on C-availability and growth status prior to amendment. It is possible that low copy number bacteria growing steadily prior to the experiment have enough ribosomal protein to support rapid and immediate growth after amendment.

# Import libraries

```{r}
library(tidyverse)
library(lmerTest)
library(emmeans)

sessionInfo()

rm(list=ls())
```

# Import and prep data

Import growth rate and start data

```{r}
# Estimated growth rates
gr.est = readRDS("rdata.files/gr.final.cleaned.rds")

# Import PAPRICA output
papEdge = read_csv("copynum_est/paprica0.5_results/growthrate.seqs.subset.bacteria.edge_data.csv")
papSeqs = read_csv("copynum_est/paprica0.5_results/growthrate.seqs.subset.bacteria.combined_16S.bacteria.tax.clean.unique.align.csv")
```

Reformat

```{r}
# Merge and remove unnecessary columns
papEdge2 = papEdge %>% select(edge_num = X1, nedge, genome_size, n16S)
papSeqs2 = papSeqs %>% select(edge_num, ASV=name, distal_length, pendant_length)
paprica = merge(papEdge2, papSeqs2, by="edge_num") %>% select(ASV, edge_num, nedge, distal_length, pendant_length, genome_size, n16S)
paprica

# Import taxonomic info
tax = read_tsv("output/growthrate.taxonomy.final.tsv") %>% select(ASV=OTU, everything())

# Merge PAPRICA results with growth rate estimates and taxonomy
paprica.gr = inner_join(paprica, gr.est, by="ASV")
paprica.gr = inner_join(paprica.gr, tax, by="ASV")

# Summarize ASVs
paprica.gr.asv = paprica.gr %>% 
  group_by(ASV, Soil, Amendment) %>% 
  summarize(k_avg = mean(k), start_avg = mean(start_day), n16S_avg=mean(n16S))
```

# Visualize

```{r}
# Overall
paprica.gr.asv %>%
  ggplot(aes(x=start_avg, y=n16S_avg)) +
  geom_point() +
  theme_test() 

# By treatment
paprica.gr.asv %>%
  ggplot(aes(x=start_avg, y=n16S_avg)) +
  geom_point() +
  facet_wrap(Soil~Amendment) +
  theme_test() 
  
```

Taxa with a wide range of 16S copy number seem to start growing immediately, whereas a few taxa with low 16S copy number started growing much later. It is possible that the low copy number taxa were already in steady growth prior to the start of the experiment, whereas high copy number taxa are more response to the resource influx at the start.

This would be a good idea to probe in the SFA experiment, where the playing ground is "even" because all taxa have to prepare for growth equally in the new environment. Pre-existing growth conditions and ribosomal loads would be less influential.

**Fit model**

This relationship looks non-linear. I think a exponential decay model would work well.

```{r}
# Self starting asymptotic model
# expression is Asym+(R0-Asym)*exp(-exp(lrc)*input)
fit <- nls(n16S_avg ~ SSasymp(start_avg, Asym, R0, lrc), data = paprica.gr.asv)
summary(fit)
```

Visualize

```{r}
# Predict values based on model
paprica.gr.asv$predict_nls = predict(fit)

# Plot with nonlinear model
paprica.gr.asv %>%
  ggplot(aes(x=start_avg, y=n16S_avg)) +
  geom_point() +
  #geom_line(aes(y=predict_nls)) +
  #geom_smooth(method="lm") +
  theme_test() 
```

This isn't going to fit.

02/10/21
After consulting with statistician, decided that best course is to not try to shoehorn this data into a particular model. Best to present data graphically and build biolgical explanation from there.

# Start day vs growth rate

```{r}
paprica.gr.asv %>%
  ggplot(aes(x=log(k_avg), y=log(start_avg))) +
  geom_point() +
  geom_smooth(method="lm") +
  facet_wrap(Soil~Amendment) +
  theme_test()
```

No clear relationships.

