---
title: "Dunlop - clean up data"
author: "Cassandra Wattenburger"
date: "2/12/2021"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(results = "show")
```

# Import libraries

```{r}
library(tidyverse)
library(phyloseq)
library(ape)

sessionInfo()

rm(list=ls())
```

# Normalized abundance data

Separate dataframes

* field data
* microcosm data

Microcosm samples

```{r}
# Import microcosm data
load("../rdata.files/gr_microcosm.norm.tax.RData")

# Clean up
ucosm <- microcosm.norm.tax %>%
  as_tibble() %>%
  select("Sample"=SampleID, Soil, "Amendment"=Cammend, "Day"=Days, "ASV"=variable, Domain, Phylum, Class, Order, Family, Genus, "norm_abund"=value) %>%
  mutate(Sample = str_remove(Sample, "sa"), Sample = as.numeric(Sample))

# Import replicate metadata
reps <- read_csv("replicate.metadata.csv")

# Add replicate information
ucosm <- ucosm %>%
  left_join(reps) %>%
  add_column(Experiment = c(rep("Microcosm", nrow(ucosm)))) %>%
  select(Experiment, Sample, Soil, Amendment, Day, "Replicate"=Rep, ASV, Domain, Phylum, Class, Order, Family, Genus, norm_abund)

ucosm
```

Field samples

```{r}
# Import field data
load("../rdata.files/gr_field.norm.tax.RData")
field <- field.norm.tax

# Clean up
field <- field %>%
  rename("Sample"=SampleID) %>%
  filter(Soil=="C3" | Soil=="S17") %>% # only relevant plots
  add_column(Experiment = c(rep("Field", nrow(.)))) %>%
  left_join(reps) %>% # replicate info
  select(Experiment, Sample, Soil, Block, "Replicate"=Rep, "ASV"=variable, Domain, Phylum, Class, Order, Family, Genus, "norm_abund"=value)

field
```

```{r, eval=FALSE}
# save
saveRDS(ucosm, file="../rdata.files/gr_ucosm.norm.clean.rds")
saveRDS(field, file="../rdata.files/gr_field.norm.clean.rds")
```

# Growth estimation data

Will include

* specific growth rate
* generation time
* window start and end day
* window start and end abundance
* 16S copy number estimated by PAPRICA
* genome size estimated by PAPRICA

**Import data**

```{r}
# Estimated growth rates/window
gr.est <- readRDS("../rdata.files/gr_gr.final_rm0.rds")

# Clean up
gr.est <- gr.est %>% 
  select(Soil, Amendment, Replicate, ASV, k, g, "start_tp"=Start, "end_tp"=End) # rename/reorder 

# Remove 0s (what was done to estimate growth rates)
ucosm.rm0 <- filter(ucosm, norm_abund > 0)

# Isolate start and end of growth data for each ASV in each treatment TROUBLESHOOT TROUBLESHOOT TROUBLESHOOT
gr.est2 = data.frame()
for (s in unique(gr.est$Soil)) {
  for (a in unique(gr.est[gr.est$Soil==s,]$Amendment)) {
    for (r in unique(gr.est[gr.est$Soil==s & gr.est$Amendment==a,]$Replicate)) {
      for (i in unique(gr.est[gr.est$Soil==s & gr.est$Amendment==a & gr.est$Replicate==r,]$ASV)) {
        gr.sub <- filter(gr.est, Soil==s & Amendment==a & Replicate==r & ASV==i)
        start.tp <- gr.sub$start_tp 
        end.tp <- gr.sub$end_tp 
        ucosm.sub <- filter(ucosm.rm0, Soil==s & Amendment==a & Replicate==r & ASV==i) %>%
          arrange(Day) # need to order by day so start and end select correct values
        start.day <- ucosm.sub[start.tp,]$Day # start day
        end.day <- ucosm.sub[end.tp,]$Day # end day
        start.abund <- ucosm.sub[ucosm.sub$Day==start.day,]$norm_abund # abundance at start of growth window
        end.abund <- ucosm.sub[ucosm.sub$Day==end.day,]$norm_abund # abundance at end of growth window
        newrow <- data.frame("Soil"=s, "Amendment"=a, "Replicate"=r, "ASV"=i,
                            "start_day"=start.day, "end_day"=end.day,
                            "start_abund"=start.abund, "end_abund"=end.abund, 
                            "k"=gr.sub$k, "g"=gr.sub$g)
        gr.est2 <- rbind(gr.est2, newrow)
      }
    }
  }
}
```

PAPRICA estimates

```{r}
# Import PAPRICA output
pap.edge <- read_csv("../paprica_analysis/paprica0.5_results/growthrate.seqs.subset.bacteria.edge_data.csv") %>% 
  select(edge_num = X1, nedge, genome_size, n16S)
pap.seqs <- read_csv("../paprika_analysis/paprica0.5_results/growthrate.seqs.subset.bacteria.combined_16S.bacteria.tax.clean.unique.align.csv") %>% 
  select(edge_num, ASV=name, distal_length, pendant_length)

# Merge
paprica <- full_join(pap.edge, pap.seqs, by="edge_num") %>% 
  select(ASV, edge_num, nedge, distal_length, pendant_length, genome_size, n16S)

# Merge PAPRICA results with growth rate data
gr.paprica <- full_join(gr.est2, paprica, by="ASV")

# taxonomy data
ucosm.tax <- select(ucosm, ASV, Domain, Phylum, Class, Order, Family, Genus) %>%
  unique()

# add
gr.paprica.tax <- left_join(gr.paprica, ucosm.tax, by="ASV")
```

```{r, eval=FALSE}
saveRDS(gr.paprica.tax, file="../rdata.files/gr_gr.paprica.clean.rds")
```

**Evaluate distal lengths**

From PAPRICA. The distal length is the phylogenetic distance from the edge where each ASV was placed on the reference tree to the furthest tip on the tree descending from that edge. We want this not to exceed 15% for the PAPRICA results to be reliable, based on results from Louca et al. 2018.

Louca, S., Doebeli, M. & Parfrey, L. W. Correcting for 16S rRNA gene copy numbers in microbiome surveys remains an unsolved problem. Microbiome 6, 41 (2018).

Numbers listed are mean, min, max, standard deviation.

Summarize distal lengths by treatment:

```{r}
paprica.distal <- gr.paprica.tax %>%
  na.omit() %>% # archaea were not estimated by PAPRICA, removes those taxa
  select(Soil, Amendment, distal_length) %>%
  group_by(Soil, Amendment) %>%
  summarize(distal_avg = mean(distal_length),
            distal_sd = sd(distal_length),
            distal_min = min(distal_length),
            distal_max = max(distal_length))
```

```{r}
paprica.distal
```

These distances look good. There may be ASVs that are much more distantly related to the reference genomes than the average, but for the most part these numbers suggest that the ASVs are well-represented by the reference genomes. Additionally, this is the distance to the furthest neighbor, not the nearest, so these numbers are over-inflated to begin with.

Interestingly, the C-amended treatments typically have lower average distal lengths than the water-amended treatments. Reference genomes had to be cultured and isolated to be sequenced and it makes sense that more "culturable" ASVs might grow in a soil treatment that was amended with a lot of simple sugars and nutrients.

# Phyloseq objects with normalized abundances

As opposed to counts.

**Microcosms**

```{r}
# Microcosm data

# Metadata
ucosm.meta <- ucosm %>%
  select(Experiment, Sample, Soil, Amendment, Day, Replicate) %>%
  unique() %>%
  mutate(rowname = Sample) %>%
  column_to_rownames(var="rowname")

# Taxonomy
ucosm.tax <- ucosm %>%
  select(ASV, Domain, Phylum, Class, Order, Family, Genus) %>%
  unique() %>%
  column_to_rownames(var="ASV") %>%
  as.matrix()

# Count
ucosm.count <- ucosm %>%
  select(Sample, ASV, norm_abund) %>%
  mutate(ASV = as.character(ASV)) %>%
  pivot_wider(names_from = "ASV", values_from = "norm_abund") %>%
  column_to_rownames(var="Sample") %>%
  as.matrix() %>%
  t()

# Tree
tree = read.tree("output/growthrate.tree.final.nwk")

# Create phyloseq object
ucosm.norm.physeq <- phyloseq(otu_table(ucosm.count, taxa_are_rows = TRUE),
                              tax_table(ucosm.tax),
                              sample_data(ucosm.meta, errorIfNULL = TRUE),
                              tree)

ucosm.norm.physeq
```

**Field**

```{r}
# Field data

# Metadata
field.meta <- field %>% 
  select(Experiment, Sample, Soil, Block, Replicate) %>%
  unique() %>%
  mutate(rowname = Sample) %>%
  column_to_rownames(var="rowname")

# Taxonomy
field.tax <- field %>%
  select(ASV, Domain, Phylum, Class, Order, Family, Genus) %>%
  unique() %>%
  remove_rownames() %>%
  column_to_rownames(var="ASV") %>%
  as.matrix()

# Count
field.count <- field %>%
  select(Sample, ASV, norm_abund) %>%
  mutate(norm_abund = as.numeric(norm_abund)) %>%
  pivot_wider(names_from = "ASV", values_from = "norm_abund") %>%
  column_to_rownames(var="Sample") %>%
  as.matrix() %>%
  t()

# Create phyloseq object
field.norm.physeq <- phyloseq(otu_table(field.count, taxa_are_rows = TRUE),
                              tax_table(field.tax),
                              sample_data(field.meta, errorIfNULL = TRUE),
                              tree)

field.norm.physeq
```

```{r, eval=FALSE}
saveRDS(ucosm.norm.physeq, "../rdata.files/gr_physeq.ucosm.norm.rds")
saveRDS(field.norm.physeq, "../rdata.files/gr_physeq.field.norm.rds")
```




